src/build/__init__.py

```diff
@@ -194,7 +74,28 @@ def _find_typo(dictionary: Mapping[str, str], expected: str) -> None:
             )
 
 
-def _parse_build_system_table(pyproject_toml: Mapping[str, Any]) -> dict[str, Any]:
+def _validate_source_directory(source_dir: PathType) -> None:
+    if not os.path.isdir(source_dir):
+        raise BuildException(f'Source {source_dir} is not a directory')
+    pyproject_toml = os.path.join(source_dir, 'pyproject.toml')
+    setup_py = os.path.join(source_dir, 'setup.py')
+    if not os.path.exists(pyproject_toml) and not os.path.exists(setup_py):
+        raise BuildException(f'Source {source_dir} does not appear to be a Python project: no pyproject.toml or setup.py')
+
+
+def _read_pyproject_toml(path: PathType) -> Mapping[str, Any]:
+    try:
+        with open(path, 'rb') as f:
+            return toml_loads(f.read().decode())
+    except FileNotFoundError:
+        return {}
+    except PermissionError as e:
+        raise BuildException(f"{e.strerror}: '{e.filename}' ")  # noqa: B904 # use raise from
```

Comment:
Can't you add `from None` or `from e` here? That's less characters than the noqa and explicitly acknowledges the chaining (which is the point of B904).

(I think you want `from None` for both of these, as they contain the required details already, not `from e`, which is the implicit default)

(Edit: Looks like this was just a moved problem from another place, rather than an added problem)

Reply:
I just factored it out from `__init__` - it was that way before :) But happy to raise `from None`.

Reply:
Fine to make that touchup in a followup, keeping moves simple is probably best. Though fine either way.

Reply:
Agree.